src/processor.ts

```diff
@@ -92,170 +118,212 @@ const processor = new SubstrateBatchProcessor()
         call: true,
         extrinsic: true,
     })
+    .includeAllBlocks()
 
 type Fields = SubstrateBatchProcessorFields<typeof processor>
-type Ctx = DataHandlerContext<Store, Fields>
+export type Call_ = Call<Fields>
+export type Event_ = Event<Fields>
+export type Extrinsic_ = Extrinsic<Fields>
+export type Block_ = Block<Fields>
+export type BlockHeader_ = BlockHeader<Fields>
+export type Ctx = DataHandlerContext<Store, Fields>
+
 export interface EventHandlerContext extends Ctx {
     block: BlockHeader<Fields>
     event: Event<Fields>
 }
 
 processor.run(new TypeormDatabase(), async (ctx) => {
-    for (let block of ctx.blocks) {
-        for (let item of block.events) {
+    // True if the most recent block is present in this batch
+    // Should always be true after syncing all blocks as each new batch will contain the last block
+    // Used to skip processing of old blocks when syncing except for genesis block
+    let isHead = ctx.isHead
+
+    // Fetch max height from the archive
+    const maxHeight = await maxHeightPromise
+
+    for (let { header: block, calls, events, extrinsics } of ctx.blocks) {
+        ctx.log.debug(
+            `block ${block.height}: extrinsics - ${extrinsics.length}, calls - ${calls.length}, events - ${events.length}`
+        )
+
+        if (isHead || block.height >= maxHeight - blockRetention) {
```

Comment:
Let's add a descriptive comment above this statement so we don't forget what we are doing here in a week or two.

Reply:
Perhaps even `isHead` is not even needed, since when that is true, the statement on the left will also be right?